sql/core/src/main/scala/org/apache/spark/sql/execution/CacheManager.scala

```diff
@@ -94,14 +94,16 @@ class CacheManager extends Logging {
       logWarning("Asked to cache already cached data.")
     } else {
       val sparkSession = query.sparkSession
-      cachedData.add(CachedData(
-        planToCache,
-        InMemoryRelation(
-          sparkSession.sessionState.conf.useCompression,
-          sparkSession.sessionState.conf.columnBatchSize,
-          storageLevel,
-          sparkSession.sessionState.executePlan(planToCache).executedPlan,
-          tableName)))
+      val inMemoryRelation = InMemoryRelation(
+        sparkSession.sessionState.conf.useCompression,
+        sparkSession.sessionState.conf.columnBatchSize,
+        storageLevel,
+        sparkSession.sessionState.executePlan(planToCache).executedPlan,
+        tableName)
+      if (planToCache.conf.cboEnabled && planToCache.stats.rowCount.isDefined) {
+        inMemoryRelation.setStatsFromCachedPlan(planToCache)
+      }
```

Comment:
I have to make InMemoryRelation stateful to avoid breaking APIs.....

Reply:
I think a more ideal change is to put the original plan stats into the constructor of `InMemoryRelation`, instead of making it mutable.

Reply:
Looks like I have no way to access InMemoryRelation from outside of spark package, though it is not a package private class...how is that achieved?

If this is the case, I can modify the constructor

Thanks @cloud-fan

Reply:
`InMemoryRelation` is not part of the public API and should be treated as unstable/internal. You can use it at your own risk. Changing the constructor is fine.

Reply:
+1